### PR TITLE
Redesign Global Fragment following Material Design Guidelines

### PR DESCRIPTION
**Pull request by ruif3r**
# Description

Redesign whole global fragment, using material cards and a pie chart, now its more easy to track the global progression of COVID. Other minor bugfixes

# ChangeType

Design changes and bugfixes

# Steps to Test

Open Global Fragment
Open Settings Activity from actionBar menu

### CODE REVIEW

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 1]
```diff
@@ -94,6 +106,37 @@ class GlobalFragment : Fragment() {
...
+        })
+    }
+
+    private fun setChartColors(): List<Int> {
+        val colors = ArrayList<Int>(3)
+        colors.add(ContextCompat.getColor(requireContext(), R.color.textColor))
+        colors.add(ContextCompat.getColor(requireContext(), R.color.recoveredColor))
+        colors.add(ContextCompat.getColor(requireContext(), R.color.deathsColor))
+        return colors
+    }
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:126]: This is not that important I will leave the comment just so you take this into account. All this methods control the chart config, maybe they could be moved to a class in charge of configuring all the pie chart specific code and configurations, this would clean up the fragment and all the responsability would be moved to this new class.

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 2]
```diff
@@ -114,5 +157,11 @@ class GlobalFragment : Fragment() {
...
         val lastUpdateLong: TextView = view.findViewById(R.id.last_update_date)
         val progressBar: ProgressBar = view.findViewById(R.id.progressBar)
+        val pieChart: PieChart = view.findViewById(R.id.circular_chart)
     }
 }
+
+fun formatLargeNumbers(value: Int): String {
+    val df = DecimalFormat("###,###,###")
+    return df.format(value)
+}
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt]: this should not be in the fragment. It is public and it doesn't have anything to do with the fragment, It should be moved to an utility formatter file. This way you could also reuse it somewhere else in your project.

File: app/src/main/res/layout/fragment_global.xml [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:fitsSystemWindows="true"
+        tools:context=".ui.global.GlobalFragment"
+        android:animateLayoutChanges="true">
+
+        <com.github.mikephil.charting.charts.PieChart
+            android:id="@+id/circular_chart"
+            android:layout_width="150dp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            android:layout_height="150dp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            android:layout_marginStart="24dp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
+
+        <TextView
+            android:id="@+id/last_update_date"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="24dp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            android:layout_marginBottom="?attr/actionBarSize"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            tools:text="Date" />
+
+        <TextView
+            android:id="@+id/last_update"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="24dp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="parent" />
+
+        <TextView
+            android:id="@+id/title"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:text="@string/global"
+            android:textSize="20sp"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: dimens

File: app/src/main/res/layout/fragment_global.xml [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:fillViewport="true"
+    android:orientation="vertical">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:fitsSystemWindows="true"
+        tools:context=".ui.global.GlobalFragment"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: do you need this context?

File: app/src/main/res/layout/fragment_global.xml [hunk 10]
```diff
@@ -0,0 +1,225 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<ScrollView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:fillViewport="true"
+    android:orientation="vertical">
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: is this orientation needed?.

File: app/src/main/res/layout/fragment_global.xml [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="@+id/title" />
+
+
+        <com.google.android.material.card.MaterialCardView
+            android:id="@+id/materialCardView"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: your Id's should follow the same convention, on the previous views you have `x_y_view_type` and now you have `xYyZz`, follow your own patterns.

File: app/src/main/res/layout/fragment_global.xml [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintStart_toStartOf="parent"
+                    tools:text="@string/percentage_example" />
+
+            </androidx.constraintlayout.widget.ConstraintLayout>
+
+        </com.google.android.material.card.MaterialCardView>
+
+        <com.google.android.material.card.MaterialCardView
+            android:id="@+id/materialCardView2"
```
Comment (jinmyjbv) [inline app/src/main/res/layout/fragment_global.xml]: in this view you have the same view structure 3 times, 

`MeterialCardView`
 -> `ConstraintLayout`
  -> `TextView`
  -> `TextView`
  -> `TextView`

as a good experiment, you could create your own custom view which contains this structure and that it receives the data that it needs to build and update itself.

**Review by jinmyjbv — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+import android.widget.TextView
+import androidx.core.content.ContextCompat
+import com.example.ncov19traking.R
+import com.google.android.material.card.MaterialCardView
+
+const val DEFAULT_TITLE_TEXT = "Title"
+const val DEFAULT_NUMBER_TEXT = "0"
+const val DEFAULT_PERCENTAGE_TEXT = "0%"
+const val CARD_ELEVATION = 10F
+const val CARD_RADIUS = 16F
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: why public?

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        percentageTextView.text = newPercentageText
+    }
+
+    private fun setupCard(){
+        setCardBackgroundColor(ContextCompat.getColor(context, R.color.backgroundColor))
+        cardElevation = CARD_ELEVATION
+        radius = CARD_RADIUS
+    }
+
+    override fun onFinishInflate() {
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: I think you don't need to wait for inflation finish, you can just create a method `initView()` and call it at the end of the `init {}` block, inflation is done almost immediately.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+                titleText = getString(R.styleable.CovidCustomCard_titleText) ?: DEFAULT_TITLE_TEXT
+                numberText = getString(R.styleable.CovidCustomCard_caseNumberText) ?: DEFAULT_NUMBER_TEXT
+                percentageText = getString(R.styleable.CovidCustomCard_percentageText) ?: DEFAULT_PERCENTAGE_TEXT
+                numberTextColor = getColor(R.styleable.CovidCustomCard_caseNumberTextColor, numberTextColor)
+                percentageTextColor = getColor(R.styleable.CovidCustomCard_percentageTextColor, percentageTextColor)
+
+            } finally {
+                recycle()
+            }
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: this process of getting variables from the styleable attributes could be abstracted to a different method.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 16]
```diff
@@ -0,0 +1,86 @@
+package com.example.ncov19traking.ui
+
+import android.content.Context
+import android.graphics.Color
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: this is not being used.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.widget.TextView
+import androidx.core.content.ContextCompat
+import com.example.ncov19traking.R
+import com.google.android.material.card.MaterialCardView
+
+const val DEFAULT_TITLE_TEXT = "Title"
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: What happens when you need internationalization, what problem will you have here? fix it.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt (hunk 13 above)
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: is this the best place to declare elevation and radius for the card?

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    private var numberText = DEFAULT_NUMBER_TEXT
+    private var percentageText = DEFAULT_PERCENTAGE_TEXT
+    private var numberTextColor = resources.getColor(R.color.recoveredColor)
+    private var percentageTextColor = resources.getColor(R.color.textColor)
+    private lateinit var titleTextView : TextView
+    private lateinit var numbersTextView : TextView
+    private lateinit var percentageTextView : TextView
+
+    init {
+        setupCard()
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: so you are setting up view configuration before the inflation here? but at the same time you are waiting for inflation for other things on the `onFinishInflate` which one is it?.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+) : MaterialCardView(context, attrs, defStyleAttr) {
+
+    private var titleText = DEFAULT_TITLE_TEXT
+    private var numberText = DEFAULT_NUMBER_TEXT
+    private var percentageText = DEFAULT_PERCENTAGE_TEXT
+    private var numberTextColor = resources.getColor(R.color.recoveredColor)
+    private var percentageTextColor = resources.getColor(R.color.textColor)
+    private lateinit var titleTextView : TextView
+    private lateinit var numbersTextView : TextView
+    private lateinit var percentageTextView : TextView
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: if you put your `findByIds` at the end of the init block you can stop using late init var here that are not bad but are expensive, and in this custom view you are not gaining anything by using them.

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 20]
```diff
@@ -64,28 +68,36 @@ class GlobalFragment : Fragment() {
...
+                        R.string.since_yesterday, homeViewModel.getCasesPercentageDifference(
+                            it.deaths,
+                            homeViewModel.nCoVYesterdayAllCases.deaths
+                        )
                     )
                 )
+                pieChart.getChartData(it)
+                pieChart.setVisibility(true)
                 progressBar.isVisible = false
             }
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:84]: this whole block could be separated in different methods to make it more readable, I see three different things being done, setting percentages, configuring piechart, and setting case numbers, try to be careful with this things too make your code more organized.

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 21]
```diff
@@ -106,13 +118,11 @@ class GlobalFragment : Fragment() {
...
-        val deaths: TextView = view.findViewById(R.id.deaths_numbers)
-        val casesPercentage: TextView = view.findViewById(R.id.percentageCasesDifference)
-        val recoveredPercentage: TextView = view.findViewById(R.id.percentageRecoveredDifference)
-        val deathsPercentage: TextView = view.findViewById(R.id.percentageDeathsDifference)
+        val covidTotalCasesCard: CovidCustomCard = view.findViewById(R.id.covid_total_cases_card)
+        val covidRecoveredCard: CovidCustomCard = view.findViewById(R.id.covid_recovered_cases_card)
+        val covidDeathsCard: CovidCustomCard = view.findViewById(R.id.covid_deaths_cases_card)
         val lastUpdateLong: TextView = view.findViewById(R.id.last_update_date)
         val progressBar: ProgressBar = view.findViewById(R.id.progressBar)
+        val pieChart = PieChartBuilder(view.findViewById(R.id.circular_chart))
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt]: mmmm I don't like the naming of this, a `PieChartBuilder` seems like a class who creates piecharts, but this class is not actually building anything is doing configurations in an specific view, you can ignore this method but is something to think about maybe next time we talk I can explain to you what I'm thinking.

File: app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt [hunk 22]
```diff
@@ -0,0 +1,10 @@
+package com.example.ncov19traking.utils
+
+import java.text.DecimalFormat
+
+class NumberFormatter
+
+fun formatLargeNumbers(value: Int): String {
+    val df = DecimalFormat("###,###,###")
+    return df.format(value)
+}
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt:11]: so the intention was to create a utility class named `NumberFormatter` but here you have a class and a method that are not related to each other.

**Review by jinmyjbv — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt (hunk 16 above)
Comment (ruif3r) [reply app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: I forgot configuring automatic optimize import

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt (hunk 19 above)
Comment (ruif3r) [reply app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: I used lateinit to have the view available for the setters. May I ask you for another way to do this? so I can put it in practice

File: app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt (hunk 22 above)
Comment (ruif3r) [reply app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt:11]: Can you explain me a bit more? shall I use companion objects?

Comment (ruif3r) [reply app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt:11]: fixed

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+    private var numberText = DEFAULT_NUMBER_TEXT
+    private var percentageText = DEFAULT_PERCENTAGE_TEXT
+    private var numberTextColor = resources.getColor(R.color.recoveredColor)
+    private var percentageTextColor = resources.getColor(R.color.textColor)
+    private var titleTextView: TextView? = null
+    private var numbersTextView: TextView? = null
+    private var percentageTextView: TextView? = null
+
+    init {
+        receivingAttributes(context, attrs, defStyleAttr)
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt]: often method naming is done in infinitive form like `receiveAttributes`, `receive`, `parse`, `get`, not `receiving`, `parsing` or `getting`

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+                    percentageText = getString(R.styleable.CovidCustomCard_percentageText)
+                        ?: DEFAULT_PERCENTAGE_TEXT
+                    numberTextColor =
+                        getColor(R.styleable.CovidCustomCard_caseNumberTextColor, numberTextColor)
+                    percentageTextColor = getColor(
+                        R.styleable.CovidCustomCard_percentageTextColor,
+                        percentageTextColor
+                    )
+
+                } finally {
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:70]: do you need this finally block here? can this throw an exception?

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+                    recycle()
+                }
+            }
+    }
+
+    private fun inflateView(context: Context) {
+        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+        inflater.inflate(R.layout.covid_custom_card, this@CovidCustomCard)
+    }
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:79]: methods ordering, private last.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+
+    fun setPercentageText(newPercentageText: String) {
+        percentageTextView?.text = newPercentageText
+    }
+
+    private fun initializeView() {
+        setCardBackgroundColor(ContextCompat.getColor(context, R.color.backgroundColor))
+        cardElevation = resources.getDimension(R.dimen.custom_covid_card_elevation)
+        radius = resources.getDimension(R.dimen.custom_covid_card_radius)
+        isClickable = true
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:85]: can't you define this clickable directly on the xml? since it seems it never changes.

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    fun setPercentageText(newPercentageText: String) {
+        percentageTextView?.text = newPercentageText
+    }
+
+    private fun initializeView() {
+        setCardBackgroundColor(ContextCompat.getColor(context, R.color.backgroundColor))
+        cardElevation = resources.getDimension(R.dimen.custom_covid_card_elevation)
+        radius = resources.getDimension(R.dimen.custom_covid_card_radius)
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:84]: nice 👍

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 28]
```diff
@@ -61,31 +67,18 @@ class GlobalFragment : Fragment() {
...
+    }
+
     private fun setupObserverSubscription() {
         homeViewModel.nCoVAllCases.observe(viewLifecycleOwner, Observer {
             with(GlobalFragmentViewHolder()) {
-                totalCases.text = it.cases.toString()
-                recovered.text = it.recovered.toString()
-                deaths.text = it.deaths.toString()
+                setCovidDataToCards(it)
+                receivingPercentageTextToCards(it)
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt]: same infinitive comment I left before

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 29]
```diff
@@ -94,9 +87,54 @@ class GlobalFragment : Fragment() {
...
+            getString(
+                R.string.since_yesterday, homeViewModel.getCasesPercentageDifference(
+                    nCoVInfo.deaths,
+                    homeViewModel.nCoVYesterdayAllCases.deaths
+                )
+            )
+        )
+    }
+
+    private fun GlobalFragmentViewHolder.setCovidDataToCards(
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt]: this name could improve, you are only calling `setCaseNumbers` so I thing it should be better to reflect that in the method name, `data` is too broad It implies you are setting different kind of models inside this method.

File: app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt [hunk 30]
```diff
@@ -106,13 +144,12 @@ class GlobalFragment : Fragment() {
...
-        val recoveredPercentage: TextView = view.findViewById(R.id.percentageRecoveredDifference)
-        val deathsPercentage: TextView = view.findViewById(R.id.percentageDeathsDifference)
+        val covidTotalCasesCard: CovidCustomCard = view.findViewById(R.id.covid_total_cases_card)
+        val covidRecoveredCard: CovidCustomCard = view.findViewById(R.id.covid_recovered_cases_card)
+        val covidDeathsCard: CovidCustomCard = view.findViewById(R.id.covid_deaths_cases_card)
         val lastUpdateLong: TextView = view.findViewById(R.id.last_update_date)
-        val progressBar: ProgressBar = view.findViewById(R.id.progressBar)
+        val progressBar: ProgressBar = view.findViewById(R.id.progress_bar)
+        val pieChart = view.findViewById<PieChart>(R.id.circular_chart)
+            .apply { PieChartConfigHelper.setupPieChart(this) }
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt]: It's better but let's better not do any kind of initialization here, you are initializing on other places of the fragment as well, why not do it outside in the on view created or where you are doing the other stuff, this seems really out of place here.

**Review by jinmyjbv — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ncov19traking/PieChartConfigHelper.kt [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+            setDrawEntryLabels(false)
+            setUsePercentValues(false)
+            setHoleColor(ContextCompat.getColor(context, R.color.backgroundColor))
+            description.isEnabled = false
+            legend.isEnabled = false
+        }
+    }
+
+    fun setAndConfigChartData(nCoVInfo: NCoVInfo, pieChart: PieChart, vararg colorList: Int) {
+        val pieEntry = ArrayList<PieEntry>()
```
Comment (jinmyjbv) [inline app/src/main/java/com/example/ncov19traking/PieChartConfigHelper.kt]: use apply.

**Review by jinmyjbv — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt (hunk 24 above)
Comment (ruif3r) [reply app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:70]: it throws a RuntimeException if the typedArray has already been recycled somewhere.

Take a look

https://developer.android.com/training/custom-views/create-view#applyattr

**Review by jinmyjbv — APPROVED** [review]
